File: Java/322. Coin Change.java
package Java;

class Solution {
  public int coinChange(int[] coins, int amount) {
    int[] dp = new int[amount + 1];
    // init
    for (int i = 0; i < dp.length; i++) {
      dp[i] = 10005;
    }

    dp[0] = 0;

    for (int i = 0; i < amount + 1; i++) {
      if (dp[i] == 10005) {
        continue;
      }
      for (int coin : coins) {
        if (i <= amount - coin) {
          dp[coin + i] = Math.min(dp[i] + 1, dp[coin + i]);
        }
      }
    }

    return dp[amount] == 10005 ? -1 : dp[amount];
  }
}

File: js/322. Coin Change.ts
function coinChange(coins: number[], amount: number): number {
  const dp = Array.from({length: amount + 1}, (value, index) => {
    return amount + 1;
  });
    dp[0] = 0;

  for(let i = 0; i < dp.length; i++) {
    if(dp[i] === amount + 1) {
       continue;
    }
    for(const coin of coins) {
      if(i <= amount - coin) {
        dp[i + coin] = Math.min(dp[i + coin], dp[i] + 1);
      }
    }
  }

  return dp[amount] === amount + 1 ? -1 : dp[amount];
};